packages/@punchcard/shape/lib/class.ts

```diff
@@ -0,0 +1,184 @@
+  public readonly Members: Members<M> = {} as any;
+
+  public readonly [Value.Tag]: I;
+
+  constructor(public readonly Type: ClassType<I, M>, public readonly Metadata: Metadata) {
+    super();
+    for (const [name, shape] of Object.entries(Type[ClassShape.Members])) {
+      (this.Members as any)[name] = new Member(name, Shape.of(shape), Meta.get(shape));
+    }
+  }
+
+  public getMetadata(): any[] {
+    return Object.values(this.Metadata);
+  }
+}
+export namespace ClassShape {
+  export type Members = typeof Members;
+  export const Members = Symbol.for('@punchcard/shape.ClassShape.Members');
+}
+
+/**
+ * A handle to a class type (the RHS of a new expression, e.g. `new RHS()`).
+ *
+ * ```ts
+ * const a = class MyClass {}
+ * typeof a; // ClassType<MyClass>;
+ * ```
+ */
+export type ClassType<I = any, M extends ClassMembers = any> =
+  (new (...args: any[]) => I) & {
+    [ClassShape.Members]: M;
+  };
+
+export type ShapeOrRecord = Shape | ClassType;
+
+export type ShapeOrRecordWithValue<T> = (Shape & { [Value.Tag]: T; }) | ClassType<T>;
+
+/**
+ * Maps ClassMembers to a structure that represents it at runtime.
+ *
+ * It supports adding `?` to optional members and maintins developer documentation.
+ *
+ * E.g.
+ * ```ts
+ * class A extends Record({
+ *   /**
+ *    * Inline documentation.
+ *    *\/
+ *   a: string
+ * })
+ *
+ * new A({
+ *   a: 'a'; // <- the above "Inline documentation" docs are preserved, traced back to the source.
+ * }).a; // <- same here
+ * ```
+ */
+export type RecordMembers<T extends ClassMembers> = {
+  /**
+   * Write each member and their documentation to the structure.
+   * Write them all as '?' for now.
+   */
+  [M in keyof T]+?: Value.Of<T[M]>;
+} & {
+  /**
+   * Remove '?' from required properties.
+   */
+  [M in RequiredKeys<T>]-?: Value.Of<T[M]>;
+};
+
+export type RecordInstance<T extends ClassMembers> = RecordMembers<T> & {
+  /**
+   * Instance reference to this record's members.
+   *
+   * Hide it with a symbol so we don't clash with the members.
+   */
+  [ClassShape.Members]: {
+    [M in keyof T]: Shape.Of<T[M]>;
+  };
+};
+
+export type RecordType<T extends ClassMembers = any> = {
+  /**
+   * Static reference to this record's members.
+   */
+  readonly [ClassShape.Members]: {
+    [M in keyof T]: Shape.Of<T[M]>;
+  };
+  /**
+   * Constructor takes values for each member.
+   */
+} & (new (values: RecordMembers<T>) => RecordInstance<T>);
+
+export function Record<T extends ClassMembers>(members: T): RecordType<T> {
+  const memberShapes: any = Object.entries(members).map(([name, member]) => ({
+    [name]: Shape.of(member)
+  })).reduce((a, b) => ({...a, ...b}));
+
+  class NewType {
+    public static readonly [ClassShape.Members] = memberShapes;
+
+    public readonly [ClassShape.Members]: T = memberShapes;
+
+    constructor(values: {
+      [K in keyof T]: Value.Of<T[K]>;
+    }) {
+      for (const [name, value] of Object.entries(values)) {
+        (this as any)[name] = value;
```

Comment:
This should only copy values for this class's properties and also validate them for optional/required/constraints etc.